File: graphs/traversals.rb
require_relative 'node'

def bfs(root, target)
  visited = []
  queue = [[root, [root]]]

  while queue.any?
    curr, path = queue.shift
    next if visited.include? curr

    visited.push(curr)

    return path.map(&:val) if curr.val == target

    curr.neighbors.each do |neighbor|
      next if visited.include?(neighbor)

      queue.push([neighbor, path + [neighbor]])
    end
  end
end

def dfs(root, target)
  visited = []
  stack = [root]

  while stack.any?
    curr = stack.pop
    visited.push(curr)

    return true if curr.val == target

    curr.neighbors.each do |neighbor|
      stack.push(neighbor) unless visited.include?(neighbor)
    end
  end
  false
end

root = Node.new(3)
node1 = Node.new(4)
node2 = Node.new(5)
node3 = Node.new(2)
node4 = Node.new(10)
node5 = Node.new(1)

root.add_edge(node1)
node1.add_edge(root)
root.add_edge(node2)
node2.add_edge(root)
root.add_edge(node3)
node3.add_edge(root)
node1.add_edge(node2)
node2.add_edge(node1)
node2.add_edge(node3)
node3.add_edge(node2)
node2.add_edge(node4)
node4.add_edge(node2)
node2.add_edge(node5)
node5.add_edge(node2)
node3.add_edge(node5)
node5.add_edge(node3)

p bfs(root, 10)
p dfs(root, 1)

File: graphs/traversals.ts
import Node from './node'

function bfs(root: Node, target: number) {
  const visited = new Set<Node>()
  const queue: Node[] = [root]

  while (queue.length > 0) {
    const current = queue.shift()
    if (current?.val === target) {
      return true
    }

    current?.neighbors.forEach((neighbor) => {
      if (!visited.has(neighbor)) {
        queue.push(neighbor)
      }
    })
  }
  return false
}

function dfs(root: Node, target: number) {
  const visited = new Set<Node>()
  const stack = [root]

  while (stack.length > 0) {
    const current = stack.pop()
    if (current?.val === target) {
      return true
    }

    current?.neighbors.forEach((neighbor) => {
      if (!visited.has(neighbor)) {
        stack.unshift(neighbor)
      }
    })
  }
  return false
}

const root = new Node(3)
const node1 = new Node(4)
const node2 = new Node(5)
const node3 = new Node(2)
const node4 = new Node(10)
const node5 = new Node(1)

root.neighbors.push(node1, node2)
node1.neighbors.push(root, node2)
node2.neighbors.push(root, node1, node3)
node3.neighbors.push(node2, node4, node5)
node4.neighbors.push(node3)
node5.neighbors.push(node3)

console.log(bfs(root, 10))
console.log(dfs(root, 10))
